### ISSUE TITLE
no type checking for @singledispatch function calls

### ISSUE THREAD

**Issue by gjcarneiro** [astral-sh/ty#2805]
### Summary

```py
from functools import singledispatch

@singledispatch
def myfunc(foo: object, bar: str) -> None:
    pass

myfunc()
myfunc(1, 2, 3)
```

No errors reported by Ty at all.   Mypy complains:

```
main.py:7: error: Missing positional arguments "foo", "bar" in call to "myfunc"  [call-arg]
main.py:8: error: Too many arguments for "myfunc"  [call-arg]
main.py:8: error: Argument 2 to "myfunc" has incompatible type "int"; expected "str"  [arg-type]
Found 3 errors in 1 file (checked 1 source file)
```

### Version

0acf998b1

**Comment by sharkdp** [comment]
Thank you for reporting this.

The typeshed stubs for `singledispatch` do not capture the full semantics of this construct. A `_SingleDispatchCallable[T]` (which is what `myfunc` turns into when decorated) can be called with arbitrary arguments according to the stubs:

https://github.com/python/typeshed/blob/da3de3ac0950f967d63a97a2d27de2ab8c1f4fa3/stubs/singledispatch/singledispatch.pyi#L16

We will need to build special support for `functools.singledispatch` if we want to support this.

**Comment by carljm** [comment]
Looks like mypy has handling for singledispatch, but pyright and pyrefly do not.

**Comment by jorenham** [comment]
I also ran into this while working on the annotations of [Gymnasium](https://github.com/Farama-Foundation/Gymnasium), which uses `@singledispatch` quite a lot (15 times at the moment: https://github.com/search?q=repo%3AFarama-Foundation%2FGymnasium%20singledispatch&type=code). Since it uses `ty` as its main type-checker, `@singledispatch` support would help a lot.
